### DISCUSSION TITLE
x-intersect only works when using both :enter and :leave

### DISCUSSION THREAD

**Discussion by fredbouin** [alpinejs/alpine#2552] [1. Help]
If I don't specifically state ```x-intersect:enter``` and ```x-intersect:leave```  all the images load at once. I am trying to use x-intersect to lazy load images and only want them to load once when entering the viewport and not disappear when leaving.

I have a grid that looks something like this (with the inside div x-data repeated many times to form the grid):

```
<div class="site-grid site-cols-12 pt-10">  
    <div x-data="{show: false}" x-intersect:enter="show = true"
         class="col-span-12 lg:col-span-4 relative overflow-hidden">    
            <div x-cloak x-show="show" 
                 x-transition:enter="transition transform ease-out duration-1000 delay-1000"
                 x-transition:enter-start="opacity-0 translate-y-1/4"
                 x-transition:enter-end="opacity-100 translate-y-0">
                <div class="h-24 bg-grey-900">...</div>
            </div>
     </div> 
</div>
```
As is all the images load at once but it works if I also have  ```x-intersect:leave="show = false"``` as stated though I don't want the leave part. I also tried doing ```<div x-intersect.once="shown = true">...</div>``` but that didn't make any difference.

**Comment by SimoTod** [comment]
can you post a codepen to replicate the issue? `x-intersect:enter` and `x-intersect:leave` do not depend on each other so what you described seems weird (There is also a unit test using `x-intersect:enter` without `x-intersect:leave` and it passes correctly).

**Reply by fredbouin** [reply]
Just tried in a codepen and everything works correctly:
https://codepen.io/fredbouin/pen/LYzdRww

It still won't work on my site though, I pushed everything to production to demonstrate:
https://captiveatelier.com/shop-all

you can see on that page the grid items just load all at once.

There are more complexities to the code on the site but locally I had it as basic as the codepen and had the same result.

**Reply by fredbouin** [reply]
Just to make things a bit simpler, I added a page on the site with the same code as the Codepen but with the problem persisting like on the shop page of the site: 

https://captiveatelier.com/test-intersect

update: I'm testing this further and what I'm seeing is when I use ```:leave``` display:none is applied to elements directly passed the viewport as expected. When I don't use leave display none only gets applied to elements that are very far down.

On the page linked above at a large breakpoint 3 column grid I don't see display none on any element. If I shrink the browser window to a 1 column layout I start seeing display none applied to the 14th element on the list.

**Reply by SimoTod** [reply]
I've just had a look and it works as expected. When you load the page, all x-show are false so all the containers have height 0 meaning that they are all in the viewport (so they all intersect and trigger the callback). if you add the x-intersect:leave, the divs going out of the viewport then trigger the leave callbacks as expected so they reapply display:none.
You need to make sure that the container have the expected height even when the content is not visible to apply your logic.

**Reply by fredbouin** [reply]
Here is an image on my end, might be easier to see, I also had the same result on a different computer. This is on initial load without having scrolled at all (without x-intersect:leave applied) see how the blue highlighted div has height 384px? That is the 14th element in the list and is definitely not in the view port. The next one down though does have height 0, so it works eventually, it just has to be way down.

<img width="1407" alt="Screen Shot 2022-01-01 at 4 02 55 PM" src="https://user-images.githubusercontent.com/14843413/147860191-8c59bd70-ad68-448e-84c6-765e0bbcc0ef.png">

**Reply by SimoTod** [reply]
You didn't read the previous answer properly. When you load the page, all your items have height = 0 because the content is hidden. You won't see it from a screenshoy because it's just a tiny fraction of time befor x-intersect starts to flip the show variables. That means that all your div stack on each other with height = 0 and are in the viewport. Hence they trigger the callback.

For example, try to put `style="height: 300 px" on all your containers having x-intersect, you'll see that only the first 1 or 2 will trigger a callback.

**Reply by fredbouin** [reply]
ahh i get what you’re saying, but as such that means that this will only work with explicit heights, no? In my project the heights of the intersect divs are defined by the height of the contained images.

**Reply by SimoTod** [reply]
Yeah, your use case is tricky, elements needs to be pushed out of the viewport before Alpine starts. If you use it to make sure you trigger the animation, then you can use the x-intersect:leave as you said.
Then you can use another `x-intersect:enter` on the image to implement the lazy loading (put the url in a data-src attribute and set src in the intersect callback) or just use `loading="lazy"' on the image tag and accept that some browsers won't support it.

**Reply by fredbouin** [reply]
I did a bit of javascript to give the div a height based off its width, this looks like it solved my problem so far.  In case anyone else runs into this type of issue:

```
    function setHeight(){
        let card = document.querySelector(".card")
        let height = card.offsetWidth * 1.33333
        card.style.height=height + 'px'
    }
    setHeight();  
    window.onresize=setHeight
```
